### DISCUSSION TITLE
编译“加速”建议，28s -> 7s（VSCode + LaTeX Workshop）

### DISCUSSION THREAD

**Discussion by Tedxz** [nju-lug/NJUThesis#218] [Show and tell]
最近在撰写博士毕业论文，目前写到约80页。按照模板手册，完整的编译流程为`xelatex -> biber -> xelatex -> xelatex`，耗时较长。以我个人的文章为例，约耗时28秒，其中`xelatex`执行一遍约为7.5秒，`biber`执行一遍约6秒。不胜其烦，探索了一下如何加速，在此跟大家分享。

首先根据相关资料，得知使用`biblatex`相较于`bibtex`可以省掉一次编译。典型的编译流程为`xelatex -> biber -> xelatex`（也适用于`pdflatex`、`latex`等）。这样大约20秒可以完成编译。

其次，当参考文献、交叉引用等内容没有更新时，根据之前编译生成的`.aux`、`.bcf`文件，只需执行一次`xelatex`即可完成编译。但是如果把默认编译流程设置为仅执行`xelatex`，更新参考文献后调用`biber`不太方便。

基于以上考虑，我把默认编译流程设置为：
`xelatex -> biber`
这样如果内容更新不多，`xelatex`执行完成后即可看到编译结果。如果更新了参考文献等，再执行一次该流程即可实现完整编译。

如果使用vscode + latex workshop插件撰写论文，可以通过配置`recipe`实现上述编译流程。插件会在整个流程结束后自动刷新内置pdf阅读器。这样大约需要等待14秒左右。

由于在大多数小改动的情况下等待`biber`执行完成并没有意义，我希望vscode能在`xelatex`执行完之后立即刷新pdf阅读器。可以通过latex workshop插件的`external build command`配置该功能。单独写一个脚本：
```sh
echo "Calling XeLaTeX on" $1
xelatex -synctex=1 -interaction=nonstopmode -file-line-error $1 
echo "\n\n\nCalling biber asyncronized on" $1
biber $1 &    
```
通过调用该脚本，`xelatex`结束后插件即认为命令结束，立即刷新pdf阅读器；`biber`在后台静默执行。如果有未处理好的引用，再次调用编译流程即可。该方法在大多数情况下等待7秒左右即可查看编译结果了。该方法存在一个已知问题，即不能在脚本返回后立即执行第二次编译，需要等`biber`运行结束，否则`biber`的结果无法参与`xelatex`编译。不过该问题也容易解决，再调用一次编译流程即可。

插件配置如下：
<img width="445" alt="image" src="https://github.com/nju-lug/NJUThesis/assets/7670294/baede972-89d1-4df2-bc13-178108d00e0f">
建议仅修改workspace设置以免弄乱其他项目的编译流程。

**Comment by note286** [comment]
用`\input`啊，注释掉暂时不关心的部分。

**Reply by Tedxz** [reply]
也可以的，不过能快速看到完整内容也不错。

**Comment by atxy-blip** [comment]
> 典型的编译流程为 `xelatex -> biber -> xelatex`

如果首次编译，我们还需要额外的一步 `xelatex` 使第三步生成的参考文献表能出现在目录里。稳妥起见就是手册中描述的四步。

办法很有参考性，赞一个。我也建议在定稿前使用 `draft` 选项，可以节省处理封面、图片和超链接的时间。

**Reply by Tedxz** [reply]
感谢大佬回复，谢谢你们的劳动成果~

在`biblatex`的[官方文档](http://mirrors.ctan.org/macros/latex/contrib/biblatex/doc/biblatex.pdf)的4.1节里提到了`biblatex`的编译流程确实可以减少一步编译，并阐述了其原因。其中157页描述了使用`bibtex`通常的四步编译流程，159页下面给出了使用`biblatex`所需的三步编译流程。我自己也测试过，在清空所有临时文件的情况下使用`xelatex -> biber -> xelatex`可以生成完整的pdf，包括目录、参考文献、交叉引用等。

我也是刚刚才了解到这一点。`biblatex`标准的三步编译流程似乎可以更新到手册里面。
